src/cicd/pipelines.ts

```diff
@@ -0,0 +1,321 @@
+}
+
+export interface AddCustomStageProps {
+  readonly stageName: string;
+  readonly steps: Step[];
+}
+
+export class CICDPipelineStack extends Stack {
+  readonly environmentId?: string;
+  readonly pipelineName?: string;
+  readonly pipelineId?: string;
+  public notificationRule?: NotificationRule;
+  public pipeline?: CodePipeline;
+  public pipelineKey?: IConstruct;
+  public sourceAction?: CodePipelineSource;
+  public synthAction?: CodeBuildStep;
+
+  constructor(
+    scope: Construct,
+    id: string,
+    environmentId: string,
+    pipelineName: string,
+    props?: StackProps,
+  ) {
+    super(scope, id, props);
+
+    this.environmentId = environmentId;
+    this.pipelineName = pipelineName;
+    this.pipelineId = id;
+  }
+
+  addSourceAction(props: SourceActionProps) {
+    var branch = props.branch ?? 'main';
+    this.sourceAction =
+      props.sourceAction ||
+      getCodeCommitSourceAction({
+        scope: this,
+        repositoryName: props.repositoryName,
+        branch: branch,
+      });
+    return this;
+  }
+
+  buildPipeline() {
+    /*
+    Build the pipeline structure.
+     Returns
+    -------
+    pipeline : CICDPipelineStack
+    CICDPipelineStack
+    */
+    this.pipeline = new CodePipeline(this, 'DDKCodePipeline', {
+      synth: this.synthAction!,
+      crossAccountKeys: true,
+      pipelineName: this.pipelineName,
+      //cliVersion: Handle when Config() is decided on
+    });
+    return this;
+  }
+
+  addSynthAction(props: SynthActionProps) {
+    this.synthAction =
+      props.synthAction ||
+      getSynthAction({
+        codePipelineSource: this.sourceAction,
+        cdkVersion: props.cdkVersion, // replace with config later on
+        partition: this.partition,
+        region: this.region,
+        account: this.account,
+        rolePolicyStatements: props.rolePolicyStatements,
+        codeartifactRepository: props.codeartifactRepository, // || this._artifactory_config.get('repository'),
+        codeartifactDomain: props.codeartifactDomain, // || this._artifactory_config.get('domain'),
+        codeartifactDomainOwner: props.codeartifactDomainOwner, //|| this._artifactory_config.get('domain_owner')
+      });
+    return this;
+  }
+
+  addStage(props: AddStageProps) {
+    /*
+    Add application stage to the CICD pipeline. This stage deploys your application infrastructure.
+      Parameters
+    ----------
+    stageId: str
+    Identifier of the stage
+    stage: Stage
+    Application stage instance
+    manualApprovals: Optional[bool]
+    Configure manual approvals. False by default
+      Returns
+    -------
+    pipeline : CICDPipelineStack
+    CICDPipelineStack
+    */
+    var manualApprovals = props.manualApprovals ?? false; // || this._config.get_env_config(stage_id).get('manual_approvals');
+
+    if (manualApprovals) {
+      this.pipeline?.addStage(props.stage, {
+        pre: [new ManualApprovalStep('PromoteTo' + toTitleCase(props.stageId))],
```

Comment:
So before this step can be run, `this.pipeline` needs to be defined as not null? We should assert this in the beginning of the function, and throw an exception if `pipeline` is not instantiated.

Reply:
Added assertion and test to validate exception.